web_src/js/markup/anchors.js

```diff
@@ -1,50 +1,67 @@
-  if (initial && document.querySelector(':target')) return;
-  if (hash?.length <= 1) return;
-  const id = decodeURIComponent(hash.substring(1));
-  const el = document.getElementById(`user-content-${id}`);
-  if (el) {
-    el.scrollIntoView();
-  } else if (id.startsWith('user-content-')) { // compat for links with old 'user-content-' prefixed hashes
+  if (!encodedId || (initial && document.querySelector(':target'))) return;
+  const id = decodeURIComponent(encodedId);
+  let el = document.getElementById(`user-content-${id}`);
+
+  // check for matching user-generated `a[name]`
+  if (!el) {
+    const nameAnchors = document.getElementsByName(`user-content-${id}`);
+    if (nameAnchors.length) {
+      el = nameAnchors[0];
+    }
+  }
+
+  // compat for links with old 'user-content-' prefixed hashes
+  if (!el && id.startsWith('user-content-')) {
     const el = document.getElementById(id);
     if (el) el.scrollIntoView();
   }
+
+  if (el) {
+    el.scrollIntoView();
+  }
 }
 
 export function initMarkupAnchors() {
-  if (!document.querySelector('.markup')) return;
-
-  // create link icons for markup headings, the resulting link href will remove `user-content-`
-  for (const heading of document.querySelectorAll(headingSelector)) {
-    const originalId = heading.id.replace(/^user-content-/, '');
-    const a = document.createElement('a');
-    a.classList.add('anchor');
-    a.setAttribute('href', `#${encodeURIComponent(originalId)}`);
-    a.innerHTML = svg('octicon-link');
-    a.addEventListener('click', (e) => {
-      scrollToAnchor(e.currentTarget.getAttribute('href'), false);
-    });
-    heading.prepend(a);
-  }
+  const markupEls = document.querySelectorAll('.markup');
+  if (!markupEls.length) return;
+
+  for (const markupEl of markupEls) {
+    // create link icons for markup headings, the resulting link href will remove `user-content-`
+    for (const heading of markupEl.querySelectorAll(`:is(h1, h2, h3, h4, h5, h6`)) {
+      const originalId = heading.id.replace(/^user-content-/, '');
+      const a = document.createElement('a');
+      a.classList.add('anchor');
+      a.setAttribute('href', `#${encodeURIComponent(originalId)}`);
+      a.innerHTML = svg('octicon-link');
+      heading.prepend(a);
+    }
+
+    // remove `user-content-` prefix from links so they don't show in url bar when clicked
+    for (const a of markupEl.querySelectorAll('a[href^="#"]')) {
+      const href = a.getAttribute('href');
+      if (!href.startsWith('#user-content-')) continue;
+      const originalId = href.replace(/^#user-content-/, '');
+      a.setAttribute('href', `#${originalId}`);
+    }
+
+    // add `user-content-` prefix to user-generated `a[name]` link targets
+    for (const a of markupEl.querySelectorAll('a[name]')) {
+      const name = a.getAttribute('name');
+      if (!name) continue;
+      a.setAttribute('name', `user-content-${a.name}`);
+    }
```

Comment:
This block is matching GitHub behaviour. Ideally we would do this prefixing in backend later, but I didn't want to touch backend here.